datanode/bootstrap/bootstrap_server.go

```diff
@@ -0,0 +1,470 @@
+package bootstrap
+
+import (
+	"context"
+	"errors"
+	pb "github.com/uber/aresdb/datanode/generated/proto/rpc"
+	"github.com/uber/aresdb/diskstore"
+	"github.com/uber/aresdb/metastore/common"
+	"github.com/uber/aresdb/utils"
+	"io"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+const (
+	chunkSize = 1024
+)
+
+var (
+	errNoCallerID       = errors.New("caller node id not set in request")
+	errNoSessionID      = errors.New("session id not set in request")
+	errInvalidSessionID = errors.New("invalid session id")
+	errInvalidRequset   = errors.New("invalid request, table/shard not match")
+	errSessionExisting  = errors.New("The request table/shard already have session running from the same node")
+)
+
+type PeerDataNodeServerImpl struct {
+	sync.RWMutex
+
+	metaStore common.MetaStore
+	diskStore diskstore.DiskStore
+
+	// session id to sessionInfo map
+	sessions map[int64]*sessionInfo
+	// tracking of all sessions for each table/shard
+	tableShardSessions map[tableShardPair][]int64
+}
+
+type tableShardPair struct {
+	table   string
+	shardID uint32
+}
+
+type sessionInfo struct {
+	sessionID    int64
+	table        string
+	shardID      uint32
+	nodeID       string
+	addr         string
+	lastLiveTime time.Time
+	ttl          int64
+}
+
+func NewPeerDataNodeServer(metaStore common.MetaStore, diskStore diskstore.DiskStore) pb.PeerDataNodeServer {
+	return &PeerDataNodeServerImpl{
+		metaStore:          metaStore,
+		diskStore:          diskStore,
+		sessions:           make(map[int64]*sessionInfo),
+		tableShardSessions: make(map[tableShardPair][]int64),
+	}
+}
+
+// IsBootstrapRunning is to check if any bootstrap is running in the table/shard
+func (p *PeerDataNodeServerImpl) IsBootstrapRunning(tableName string, shardID uint32) bool {
+	p.RLock()
+	defer p.RUnlock()
+
+	_, ok := p.tableShardSessions[tableShardPair{table: tableName, shardID: shardID}]
+	if !ok {
+		return true
+	}
+	return false
+}
+
+// StartSession create new session for one table/shard/node, only One session can be established on one table/shard from one node
+func (p *PeerDataNodeServerImpl) StartSession(ctx context.Context, req *pb.StartSessionRequest) (*pb.Session, error) {
+	var err error
+	sessionInfo := &sessionInfo{
+		table:        req.Table,
+		shardID:      req.Shard,
+		nodeID:       req.NodeID,
+		ttl:          req.Ttl,
+		lastLiveTime: utils.Now(),
+		sessionID:    rand.Int63(),
```

Comment:
monotonically increasing number might be better for the id?

Reply:
how about some combination of table shard node and sequence number

Reply:
hmm, that works too local id generator, will change